bpo-14678: Update zipimport to support importlib.invalidate_caches()

desmondcheongzx opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->

Added an invalidate_caches() method to the zipimport.zipimporter class based on the implementation of importlib.FileFinder.invalidate_caches(). This was done by adding a get_files() method and an _archive_mtime attribute to zipimport.zipimporter to check for updates or cache invalidation whenever the cache of files and toc entry information in the zipimporter is accessed.

<!-- issue-number: [bpo-14678](https://bugs.python.org/issue14678) -->
https://bugs.python.org/issue14678
<!-- /issue-number -->

Automerge-Triggered-By: GH:brettcannon

desmondcheongzx committed:
Add a get_files() method to zipimporter

Functions in zipimport and zipimporter methods used to access the
files and toc entries stored in a ZIP archive via zipimporter's
private _files attribute. However, this makes it hard to check
if the ZIP archive had been updated, or if the cached information
has been invalidated whenever _files is accessed.

Hence, this change adds get_files() as an internal accessor for the
_files attribute. Whenever get_files() is called, we do a check to
make sure that our cached information is up-to-date. This check is
done using a new _archive_mtime attribute that stores the last
modified timestamp of the ZIP archive whenever
zipimport._read_directory is called.

desmondcheongzx committed:
Add an invalidate_caches() method to zipimporter

Zipimport's zipimporter are the finders that get cached in
sys.path_importer_cache, so this addition enables zipimport to
support importlib.invalidate_caches().

desmondcheongzx committed:
Add a news entry

desmondcheongzx committed:
Add an additional test for zipimporter.invalidate_caches()

desmondcheongzx committed:
Rename get_files() to _get_files()

brettcannon commented on Lib/zipimport.py:
```diff
@@ -320,6 +321,29 @@ def get_resource_reader(self, fullname):
+
+    def invalidate_caches(self):
+        """Invalidate the mtime of the ZIP archive."""
+        self._archive_mtime = -1
```
Is there a specific reason to add the stat call on every access of the importer instead of keeping the cache as-is and instead rebuilding it when `invalidate_caches()` is called (i.e. make `invalidate_caches()` take over the part of `__init__()` that sets `self._files`)? Since the common case is going to be no file is changed, I think optimizing performance that way to avoid the stat call makes sense.

You also didn't remove `self._files`, which led to two different sets of file data.

desmondcheongzx replied:
In this case I was following the pattern in `FileFinder.find_spec()` where the cache is refreshed whenever the directory that a FileFinder is handling has been modified. Maybe it isn't so applicable because a zip file isn't modified as often.

If we're removing this check for zip file modification and consolidating setting `self._files` in `invalidate_caches()`, then it also makes sense to revert `_get_files()` back to `self._files` since no additional checks are being done on accessing file data. Keeping track of the zip file's mtime probably isn't needed anymore either since a user presumably always wants to call `_read_directory()` if `invalidate_caches()` is being called.

brettcannon reviewed (changes requested):
I think it would make more sense to consolidate that code that sets `self._files` in `invalidate_caches()` and call it from `__init__()` to avoid having to make so many stat calls.

desmondcheongzx committed:
Consolidate updates of _files into invalidate_caches()

This change also reverts the new _get_files() method back to the
_files private attribute, and removes the addition of the
_archive_mtime attribute.

desmondcheongzx committed:
Remove stray newline

brettcannon commented on Misc/NEWS.d/next/Library/2021-01-07-21-25-49.bpo-14678.1zniCH.rst:
```diff
@@ -0,0 +1,2 @@
+Add an invalidate_caches() method to the zipimport.zipimporter class to
+support importlib.invalidate_caches().
```
```suggestion
support importlib.invalidate_caches().
Patch by Desmond Cheong.
```

desmondcheongzx committed:
Add author to news entry

desmondcheongzx committed:
Merge remote-tracking branch 'upstream/master' into bpo-14678-add-invalidate-caches-to-zipimporter

jaraco commented on Lib/test/test_zipimport.py:
```diff
@@ -506,6 +506,32 @@ def testZipImporterMethods(self):
         self.assertEqual(zi2.archive, TEMP_ZIP)
         self.assertEqual(zi2.prefix, TESTPACK + os.sep)
 
+        # Test invalidate_caches() method
```
This method is already egregiously long. Would you consider creating a separate method for it?

jaraco reviewed (approved):
One minor nitpick, but otherwise, looks good to me.

desmondcheongzx committed:
Move tests for invalidate_caches into a new method

desmondcheongzx committed:
Update bytecode for importlib._zipimport.h

HyukjinKwon commented on Lib/zipimport.py:
```diff
@@ -321,6 +321,16 @@ def get_resource_reader(self, fullname):
+    def invalidate_caches(self):
+        """Reload the file data of the archive path."""
+        try:
+            self._files = _read_directory(self.archive)
```
Thanks, @desmondcheongzx. I do appreciate this fix.

Just dropping a comment in case the impact of this was missed during the review.
This might impact the performance notably. For example, I only import two zipped packages `py4j` and `pyspark` (and of course I have many packages installed from pip, and import them within my code) but it slows down the speed of `importlib.invalidate_caches` up to 3500% because now it has to go through the zipped package.

desmondcheongzx replied:
This was missed unfortunately.

To get everyone back up to speed, in the [original discussion](https://github.com/python/cpython/pull/24159#discussion_r571303809) there were two decisions made:
1.  The `_get_files()` method should be as fast as possible in the common case where zip files are not changed. This is why we avoided a stat call on every access of the importer.
2. Rebuild the cache when `invalidate_caches()` is called. This second decision might not be the right one, because clearing a cache does not mean that we want to immediately repopulate it.

There are three relevant events:
1. `_get_files()` is called with a valid cache
2. `invalidate_caches` is called
3. `_get_files()` is called with an __invalid__ cache

IMO, what we really want is for events (1) and (2) to be fast, while event (3) can be slow since we're repopulating a cache. (Note that in the original PR we made (1) and (3) fast, but (2) slow). So I propose the following:

- Add a boolean flag `cache_is_valid` that is set to false when `invalidate_caches()` is called.
- In `_get_files()`, if `cache_is_valid` is true, use the cache. If `valid_cache` is false, call `_read_directory()`.

This approach avoids any behaviour change and keeps the common path performant, while also shifting the cost of reading the directory out of cache invalidation.

If we want to go further and consider the fact that we rarely expect zip archives to change, then we can also consider adding a flag to `importlib.invalidate_caches` so that users can choose whether zip caches are invalidated.

Any thoughts, @brettcannon?
